chore: add migration guide for v5 (#257)
Unleash/unleash-client-dotnet: v5_MIGRATION_GUIDE.md

# Migrating to Unleash-Client-Dotnet 5.0.0

This guide highlights the key changes you should be aware of when upgrading to v5.0.0 of the Unleash client.

## Custom strategy changes

Custom strategies no longer provide the option to access the constraints in their interface. The method `bool IsEnabled(Dictionary<string, string> parameters, UnleashContext context, IEnumerable<Constraint> constraints)` no longer exists. Other than that, custom strategies remain unchanged.

> ## Direct access to feature toggles

Reviewer: We also removed `GetVariants`

Author: Good spot, thank you!

Author: Made a note in [last commit](https://github.com/Unleash/unleash-client-dotnet/pull/257/commits/b199365e9352c293b7b128c5fe863bc5b8073e66)

> Direct access to the feature toggle objects through `UnleashClient.FeatureToggles` has been removed. All classes related to the internal representation of feature toggles is no longer publicly accessible in the SDK.

Reviewer: > All classes related to the internal representation of feature toggles is no longer publicly accessible in the SDK.

```suggestion
Direct access to the feature toggle objects through `UnleashClient.FeatureToggles` has been removed. All classes related to the internal representation of feature toggles are no longer publicly accessible in the SDK.
```
?
Maybe I can't brain this properly

Author: No no, I think me no brain good. You brain good

# Migrating to Unleash-Client-Dotnet 5.0.0

This guide highlights the key changes you should be aware of when upgrading to v5.0.0 of the Unleash client.

## Custom strategy changes

Custom strategies no longer provide the option to access the constraints in their interface. The method `bool IsEnabled(Dictionary<string, string> parameters, UnleashContext context, IEnumerable<Constraint> constraints)` no longer exists. Other than that, custom strategies remain unchanged.

## Direct access to feature toggles

Direct access to the feature toggle objects through `UnleashClient.FeatureToggles` has been removed. All classes related to the internal representation of feature toggles are no longer publicly accessible in the SDK.

The SDK now provides a `UnleashClient.ListKnownToggles` method, which will return a list of feature toggle names, their type and the project they're bound to.

> The client also no longer provides access to list the variants bound to a feature toggle through `UnleashClient.GetVariants`. We believe this was inappropriately leaking the internal abstractions out of the SDK but if you have a strong use case for this, please open an issue.

Reviewer 2: ```suggestion
The client also no longer provides access to listing the variants bound to a feature flag through `UnleashClient.GetVariants`. We determined that this was exposing internal abstractions that should remain within the SDK. However, if you have a strong use case for this, please open an issue.
```

## Bootstrapping changes

Due to the changes in the previous section, Bootstrapping classes are now required to return a `String` instead of a `FeatureToggleCollection`. The string should be a JSON string representing the response returned from your Unleash instance's api/client/features endpoint. In practice, that means if you previously had a `Read` method in your bootstrapping class like so:

``` dotnet

public ToggleCollection Read()
{
    var json = settings.FileSystem.ReadAllText(filePath);
    return settings.JsonSerializer.Deserialize<ToggleCollection>(json);
}

```

This can now be simplified to:

``` dotnet

public string Read()
{
    return settings.FileSystem.ReadAllText(filePath);
}

```

## Custom serializers

> In v4.x and before, the SDK provided the option to use NewtonSoft or mount a custom JSON serializer. This option has been removed in v5.x, the SDK now relies on System.Text.Json and no option to override that is provided. If you were previously providing a custom serializer to access System.Text.Json, that's now safe to remove. Users who were previously relying on NewtonSoft shouldn't have to make any changes.

Reviewer 2: Since we still support NewtonSoft, I find the sentence 'In v4.x and before, the SDK provided the option to use NewtonSoft' misleading. If you keep reading, you find out that it's OK to use, but it confused me for a second. Can we just skip it?

```suggestion
In v4.x and before, the SDK provided the option of mounting a custom JSON serializer. This option has been removed in v5.x, the SDK now relies on `System.Text.Json` with no option to override it. If you previously provided a custom serializer to access `System.Text.Json`, it's now safe to remove it.

If you use NewtonSoft you don't have to make any changes.
```

Reviewer: I'll add some context here. Some of the code we're removing with this release had support for auto wiring up newtonsoft as the serializer if available. The alternative to using Newtonsoft with auto wiring was to write some plumbing code to integrate that other json serializer (which seems to usually have been System.Text.Json). That has all been replaced with a hard dependency on System.Text.Json and the interfaces for custom plumbing code is gone.

So this text tackles two things. That the people who needed to do zero changes due to using Newtonsoft can expect that nothing breaks but we're not using Newtonsoft for serialization, and if you did make a custom serializer then that's a breaking change for you

Don't know if this changes what you think would be good description for that here, just wanted to provide a little context :)

Author: Pretty much what David said! If you're using NewtonSoft for other things in your code, then you won't have to worry. But no NewtonSoft code will now touch the SDK code. Open to suggestions if you think this isn't clear?
